torch/autograd/gradcheck.py

```diff
@@ -871,8 +871,61 @@ def _test_backward_mul_by_grad_output(outputs, inputs, check_sparse_nnz) -> bool
             raise GradcheckError('grad is incorrect size')
     return True
 
+def _test_undefined_forward_mode(func, outputs, inputs):
+    fwAD = torch.autograd.forward_ad
+
+    inp_tensors_idx, inp_tensors = _get_inp_tensors(inputs)
+    all_v, all_u, all_u_dense = _make_vectors(inp_tensors, outputs, use_forward_ad=True)
+
+    tensor_inputs = tuple(i for i in inputs if is_tensor_like(i) and i.requires_grad)
+
+    with fwAD.dual_level():
+        fw_grads = []
+        dual_inputs = []
+        for i, inp in enumerate(inputs):
+            if is_tensor_like(inp) and inp.requires_grad:
+                if inp.layout == torch._mkldnn:  # type: ignore[attr-defined]
+                    raise ValueError("MKLDNN inputs are not support for forward AD gradcheck.")
+
+                inp = fwAD.make_dual(inp, torch.zeros_like(inp))
+                # If inp is a differentiable view, the dual might not be the tangent given to
+                # make_dual, so read it explicitly from the dual tensor
+                fw_grads.append(fwAD.unpack_dual(inp)[1])
+            dual_inputs.append(inp)
+
+        for i, (fw_grad, u) in enumerate(zip(fw_grads, all_u)):
+            fw_grad.copy_(u.view_as(fw_grad))
+
+        dual_inputs_idx = 0
+        for inp in inputs:
+            if is_tensor_like(inp) and inp.requires_grad:
+                dual_inp_obj = dual_inputs[dual_inputs_idx]
+
+                # case 1 (Materialized Zero Tensor Tangent)
+                dual_inputs[dual_inputs_idx] = fwAD.make_dual(inp, torch.zeros_like(inp))
+                raw_outputs = _as_tuple(func(*dual_inputs))
+                dual_outputs1 = filter(_is_float_or_complex_tensor, raw_outputs)
+
+                # case 2 (Efficient Zero Tensor Tangent since we don't make a dual object and pass a regular tensor)
+                dual_inputs[dual_inputs_idx] = inp
+                raw_outputs = _as_tuple(func(*dual_inputs))
+                dual_outputs2 = filter(_is_float_or_complex_tensor, raw_outputs)
+
+                # reset
+                dual_inputs[dual_inputs_idx] = dual_inp_obj
+                dual_inputs_idx += 1
+
+                for index_o, (d_o1, d_o2) in enumerate(zip(dual_outputs1, dual_outputs2)):
+                    val1, res1 = fwAD.unpack_dual(d_o1)
+                    val2, res2 = fwAD.unpack_dual(d_o2)
+
+                    if not (res1 is None or res2 is None):
```

Comment:
nit check that val1/val2 match as well?

Reply:
that check doesn't need to be in this test